internal/pagination/pagination2.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+// ListResponse2 represents the response from the paginated list operation.
+type ListResponse2[T boundary.Resource] struct {
+	Items              []T
+	CompleteListing    bool
+	RefreshToken       *refreshtoken.Token
+	DeletedIds         []string
+	EstimatedItemCount int
+}
+
+// ListFilterFunc is a callback passed down from the application service layer
+// used to filter out resources that don't match any user-supplied filter.
+type ListFilterFunc[T boundary.Resource] func(ctx context.Context, item T) (bool, error)
+
+// ListItemsFunc2 is used to encapsulate the creation of repository specific options
+// and listing of items.
+type ListItemsFunc2[T boundary.Resource] func(ctx context.Context, prevPageLastItem T, limit int) ([]T, error)
+
+// EstimatedCountFunc is used to estimate the total number of items
+// available in the table given any filter provided.
+type EstimatedCountFunc func(ctx context.Context) (int, error)
+
+// ListDeletedIDsFunc is used to list the IDs of the resources deleted since
+// the given timestamp. It returns a slice of IDs and the timestamp of the
+// instant in which the slice was created.
+type ListDeletedIDsFunc func(ctx context.Context, since time.Time) ([]string, time.Time, error)
+
+// List lists items according to the page size,
+// filtering out entries that do not
+// pass the filter item fn. It returns a new refresh token
+// based on the grants hash and the returned items.
+func List[T boundary.Resource](
+	ctx context.Context,
+	grantsHash []byte,
+	pageSize int,
+	filterItemFn ListFilterFunc[T],
+	listItemsFn ListItemsFunc2[T],
+	estimatedCountFn EstimatedCountFunc,
+) (*ListResponse2[T], error) {
+	const op = "pagination.List"
+
+	items, completeListing, err := list(ctx, grantsHash, pageSize, filterItemFn, listItemsFn)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	resp := &ListResponse2[T]{
+		Items:              items,
+		CompleteListing:    completeListing,
+		EstimatedItemCount: len(items),
+	}
+
+	if !completeListing {
+		// If this was not a complete listing, get an estimate
+		// of the total items from the DB.
+		var err error
+		resp.EstimatedItemCount, err = estimatedCountFn(ctx)
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+	}
+
+	if len(items) > 0 {
+		resp.RefreshToken = refreshtoken.FromResource(items[len(items)-1], grantsHash)
+	}
+
+	return resp, nil
+}
+
+// ListRefresh lists items according to the page size
+// and refresh token, filtering out entries that do not
+// pass the filter item fn. It returns a new refresh token
+// based on the old one, the grants hash, and the returned
+// items.
+func ListRefresh[T boundary.Resource](
+	ctx context.Context,
+	tok *refreshtoken.Token,
+	grantsHash []byte,
+	pageSize int,
+	filterItemFn ListFilterFunc[T],
+	listItemsFn ListItemsFunc2[T],
+	estimatedCountFn EstimatedCountFunc,
+	listDeletedIDsFn ListDeletedIDsFunc,
```

Comment:
I am wondering if there is a way to ensure that these functions only operate on the resource `T`. As it stands, there is nothing at compile time that would prevent using `target.Target` for `T` and then passing an `estimatedCountFn` that provides the count of `hosts.Host`.

Reply:
Good point. As you recall, originally this functionality was behind a single repository interface rather than 3 separate callbacks. I wonder if it would make sense to add that single repository back, so that the caller is forced to have
```
ListItems(..)([]T, error)
EstiamtedCount(...)(int, error)
ListDeletedIds(...)([]string, time.Time, error)
```
On a single value? It would add a bit of toil on the caller side because you'd have to have some type that just curries the functions but it would at least tie the `T` to the other two functions, perhaps?

Alternatively, having them take a `T` parameter that's completely unused, but that seems like it could get confusing for future readers. What do you think?

Reply:
I am not sure I like either enough to request a specific change.